cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -405,14 +408,38 @@ Status CopyStridedArray(PyArrayObject* arr, const int64_t length, MemoryPool* po
 }  // namespace
 
 template <typename ArrowType>
-inline Status NumPyConverter::ConvertData(std::shared_ptr<Buffer>* data) {
+inline Status NumPyConverter::PrepareInputData(std::shared_ptr<Buffer>* data) {
   if (is_strided()) {
     RETURN_NOT_OK(CopyStridedArray<ArrowType>(arr_, length_, pool_, data));
+  } else if (dtype_->type_num == NPY_BOOL) {
+    int64_t nbytes = BitUtil::BytesForBits(length_);
+    std::shared_ptr<Buffer> buffer;
+    RETURN_NOT_OK(AllocateBuffer(pool_, nbytes, &buffer));
+
+    Ndarray1DIndexer<uint8_t> values(arr_);
+
+    uint8_t* bitmap = buffer->mutable_data();
+
+    memset(bitmap, 0, nbytes);
+    for (int i = 0; i < length_; ++i) {
+      if (values[i] > 0) {
+        BitUtil::SetBit(bitmap, i);
```

Comment:
This loop would probably be faster if replaced with `GenerateBits` (see `bit-util.h`).

Reply:
Yes, let's definitely do that

Reply:
Ok, I'll give that a shot